### PR TITLE
*: upgrade go1.21

### LINKED ISSUE #45925: upgrade go1.21
## Enhancement

🥳 Go 1.21.0 is released!.

Release notes: [https://go.dev/doc/go1.21](https://t.co/8yX6FMmTOq)

so we need to upgrade for tidb.

1、 Upgrade ci image https://github.com/PingCAP-QE/ci/pull/2341
2、 Set bazel go1.21.0 https://github.com/pingcap/tidb/pull/45923

### PR DESCRIPTION
**Pull request by hawkingrei**
### What problem does this PR solve?

Issue Number: close #45925

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by lance6716 — APPROVED** [review]

**Review by Defined2014 — APPROVED** [review]

File: tools/patch-go/build.sh [hunk 1]
```diff
@@ -16,7 +16,7 @@
 set -xeuo pipefail
 
 # When updating to a new Go version, update all of these variables.
-GOVERS=1.20.5
+GOVERS=1.21.0
 GOLINK=https://go.dev/dl/go$GOVERS.src.tar.gz
 SRCSHASUM=9a15c133ba2cfafe79652f4815b62e7cfc267f68df1b9454c6ab2a3ca8b96a88
```
Comment (wuhuizuo) [inline tools/patch-go/build.sh:21]: It's not updated, please update it.

**Pull request merged**